pkg/volume/azure_dd/attacher.go

```diff
@@ -0,0 +1,289 @@
+	host          volume.VolumeHost
+	azureProvider azureCloudProvider
+}
+
+var _ volume.Attacher = &azureDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &azureDataDiskPlugin{}
+
+const (
+	checkSleepDuration = time.Second
+)
+
+// Azure doesn't allow simultanious disk attach or detach on the same instance
+// serialize concurrent attach/detach per instance
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *azureDataDiskPlugin) NewAttacher() (volume.Attacher, error) {
+	azure, err := getAzureCloudProvider(plugin.host.GetCloudProvider())
+	if err != nil {
+		glog.V(4).Infof("failed to get azure provider")
+		return nil, err
+	}
+
+	return &azureDiskAttacher{
+		host:          plugin.host,
+		azureProvider: azure,
+	}, nil
+}
+
+func (attacher *azureDiskAttacher) Attach(spec *volume.Spec, hostName string) (string, error) {
+	volumeSource, err := getVolumeSource(spec)
+	if err != nil {
+		glog.V(2).Infof("failed to get azure disk spec")
+		return "", err
+	}
+	instanceid, err := attacher.azureProvider.InstanceID(hostName)
+	if err != nil {
+		glog.V(2).Infof("failed to get azure instance id")
+		return "", fmt.Errorf("failed to get azure instance id for host %q", hostName)
+	}
+	if ind := strings.LastIndex(instanceid, "/"); ind >= 0 {
+		instanceid = instanceid[(ind + 1):]
+	}
+	attachDetachMutex.LockKey(instanceid)
+	defer attachDetachMutex.UnlockKey(instanceid)
+
+	lun, err := attacher.azureProvider.GetDiskLun(volumeSource.DiskName, volumeSource.DataDiskURI, instanceid)
+	if err == cloudprovider.InstanceNotFound {
+		// Log error and continue with attach
+		glog.Warningf(
+			"Error checking if volume is already attached to current node (%q). Will continue and try attach anyway. err=%v",
+			instanceid, err)
+	}
+
+	if err == nil {
+		// Volume is already attached to node.
+		glog.V(4).Infof("Attach operation is successful. volume %q is already attached to node %q at lun %d.", volumeSource.DiskName, instanceid, lun)
+	} else {
+		lun, err = attacher.azureProvider.GetNextDiskLun(instanceid)
+		if err != nil {
+			glog.Warningf("no LUN available for instance %q", instanceid)
+			return "", fmt.Errorf("all LUNs are used, cannot attach volume %q to instance %q", volumeSource.DiskName, instanceid)
+		}
+
+		err = attacher.azureProvider.AttachDisk(volumeSource.DiskName, volumeSource.DataDiskURI, instanceid, lun, compute.CachingTypes(*volumeSource.CachingMode))
+		if err == nil {
+			glog.V(4).Infof("Attach operation successful: volume %q attached to node %q.", volumeSource.DataDiskURI, instanceid)
+		} else {
+			glog.V(2).Infof("Attach volume %q to instance %q failed with %v", volumeSource.DataDiskURI, instanceid, err)
+			return "", fmt.Errorf("Attach volume %q to instance %q failed with %v", volumeSource.DiskName, instanceid, err)
+		}
+	}
+
+	return strconv.Itoa(int(lun)), err
+}
+
+func (attacher *azureDiskAttacher) WaitForAttach(spec *volume.Spec, lunStr string, timeout time.Duration) (string, error) {
```

Comment:
It's not an object, it is very specifically a field with specific meaning and used as such. I suggest leave it empty on the Attach if you must, and having WaitForAttach figure it out and return it correctly (which it appears to already do). If the only way the node can figure out what the path is is using the LUN and the LUN is returned on attach, then you will need to first make the bigger change to the underlying infrastructure to allow device path to be an opaque meaningless string.